lib/codegen/CreateTerm.cpp

```diff
@@ -99,100 +101,334 @@ llvm::Type *getValueType(SortCategory sort, llvm::Module *Module) {
   case SortCategory::Set:
     return Module->getTypeByName(SET_STRUCT);
   case SortCategory::Int:
-    return llvm::PointerType::get(Module->getTypeByName(INT_STRUCT), 0);
+    return llvm::PointerType::getUnqual(Module->getTypeByName(INT_STRUCT));
   case SortCategory::Float:
-    return llvm::PointerType::get(Module->getTypeByName(FLOAT_STRUCT), 0);
+    return llvm::PointerType::getUnqual(Module->getTypeByName(FLOAT_STRUCT));
   case SortCategory::StringBuffer:
-    return llvm::PointerType::get(llvm::PointerType::get(Module->getTypeByName(STRING_STRUCT), 0), 0);
+    return llvm::PointerType::getUnqual(llvm::PointerType::getUnqual(Module->getTypeByName(STRING_STRUCT)));
   case SortCategory::Bool:
-    return llvm::IntegerType::get(Module->getContext(), 1);
+    return llvm::Type::getInt1Ty(Module->getContext());
   case SortCategory::MInt:
     assert(false && "not implemented yet: MInt");
   case SortCategory::Symbol:
-    return llvm::PointerType::get(Module->getTypeByName(BLOCK_STRUCT), 0);
+    return llvm::PointerType::getUnqual(Module->getTypeByName(BLOCK_STRUCT));
   }
 }
 
 
-llvm::StructType *getBlockType(llvm::Module *Module, const KOREObjectSymbol *symbol, KOREObjectSymbolDeclaration *symbolDecl) {
+llvm::StructType *getBlockType(llvm::Module *Module, KOREDefinition *definition, const KOREObjectSymbol *symbol, KOREObjectSymbolDeclaration *symbolDecl) {
   llvm::StructType *BlockHeaderType = Module->getTypeByName(BLOCKHEADER_STRUCT);
-  llvm::ArrayType *EmptyArrayType = llvm::ArrayType::get(llvm::IntegerType::get(Module->getContext(), 64), 0);
+  llvm::ArrayType *EmptyArrayType = llvm::ArrayType::get(llvm::Type::getInt64Ty(Module->getContext()), 0);
   llvm::SmallVector<llvm::Type *, 4> Types;
   Types.push_back(BlockHeaderType);
   Types.push_back(EmptyArrayType);
   for (KOREObjectSort *arg : symbol->getArguments()) {
     auto sort = dynamic_cast<KOREObjectCompositeSort *>(arg);
-    llvm::Type *type = getValueType(sort->getCategory(), Module);
+    llvm::Type *type = getValueType(sort->getCategory(definition), Module);
     Types.push_back(type);
   }
   return llvm::StructType::get(Module->getContext(), Types);
 }
 
-llvm::Value *getBlockHeader(llvm::Module *Module, const KOREObjectSymbol *symbol) {
+llvm::Value *getBlockHeader(llvm::Module *Module, KOREDefinition *definition, const KOREObjectSymbol *symbol, llvm::Type *BlockType) {
   llvm::StructType *BlockHeaderType = Module->getTypeByName(BLOCKHEADER_STRUCT);
   uint64_t headerVal = symbol->getTag();
-  headerVal |= (uint64_t)symbol->length() << 32;
+  headerVal |= (llvm::DataLayout(Module).getTypeAllocSize(BlockType) / 8) << 32;
```

Comment:
Truncating division looks a bit suspicious here. Maybe have a method return the block count, or round up here.

Reply:
I'm assuming that getTypeAllocSize returns a multiple of 8 because BlockType is guaranteed to be at least 64-bit aligned, but I will add an explicit assertion